sycl/include/CL/sycl/detail/spirv.hpp

```diff
@@ -122,9 +122,12 @@ AtomicCompareExchange(multi_ptr<T, AddressSpace> MPtr,
   auto SPIRVSuccess = getMemorySemanticsMask(Success);
   auto SPIRVFailure = getMemorySemanticsMask(Failure);
   auto SPIRVScope = getScope(Scope);
-  auto *Ptr = MPtr.get();
-  return __spirv_AtomicCompareExchange(Ptr, SPIRVScope, SPIRVSuccess,
-                                       SPIRVFailure, Desired, Expected);
+  using I = detail::atomic_integer_arg_t<T>;
+  auto *Ptr = reinterpret_cast<typename multi_ptr<I, AddressSpace>::pointer_t>(
+      MPtr.get());
+
+  return (T)__spirv_AtomicCompareExchange(
+      Ptr, SPIRVScope, SPIRVSuccess, SPIRVFailure, (I)Desired, (I)Expected);
```

Comment:
Are these changes necessary?  The `enable_if` should already limit `T` to integral types, and so it seems like `detail::atomic_integer_arg_t<T>` will only ever return `T` itself.

Same question for all the functions in this file.

Reply:
This may require a bit of background. `clang/lib/Sema/SPIRVBuiltins.td` normalizes the builtin definitions an implementation can produce. Because it only exposes valid SPIR-V builtins, it protects the programmer. This all fits into the SPIR-V ecosystem - it represents a set of builtins that are usable in OpenCL and allows its implementation as a library. This means that projects other than SYCL can make use of the builtins, which will be good for SPIR-V and upstreaming. 

Because SYCL's atomic code is templated, it is possible that the user will attempt to use integer types that are not have corresponding SPIR-V atomic builtins. SPIR-V atomics do not provide implementations for `long long` or `unsigned long long` arguments, even though in practice they may be the same size as `long` or `unsigned long`. To work around this, we circumvent the type system so that `long long` and `unsigned long long` can be passed from SYCL to the spriv builtins. This PR introduces some machinery to pick an appropriately sized integer type and performs the casting as needed.

Reply:
Ah, I see.  That makes sense -- thanks for the explanation.

Could we put a comment above the atomic functions (just one, before `AtomicCompareExchange`) that says something to this effect?  Something as simple as the below would do:
```
// C++ integer types are converted for SPIR-V built-ins
// e.g. unsigned long long => unsigned long
```